Bluetooth: Mesh: Refactor Mesh Model Extensions

michal-narajowski opened the pull request:
The existing extension tree does not support all the features that are
defined by the specification (e.g. multiple parents).

To enable full support for Model Extensions this patch introduces graph
representation in a form of edge list. This representation allows for
fine-grained control over memory usage. The downside is that traversing
the graph is sub-optimal computationally.

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

trond-snekvik reviewed (changes requested):
You were right about that graph_walk issue, that's a pretty heavy process. 

I like the fact that models that don't extend anything don't consume any memory, but I think there are a couple of issues here that I would like to resolve:

1. graph_walk is non-reentrant, slow and because of the recursion, its stack depth is unknown at compile time. We're walking the graph of every model on every received group message, so the processing time is significant.
2. Users have to calculate the number of extensions their composition data has (or look it up at runtime). This is not easy to determine, and we have successfully avoided this type of configuration entries until now.
3. The extension entries are pointing to both models in the relationship, so every relationship is 8 bytes. Feels like this could be reduced somehow.

Point 1 is definitely the biggest issue here, and although we agreed that performance could be fixed later, the additional stack depth and non-reentrant properties of this approach, combined with the lack of a clear solution to this issue in this approach makes me want to address this right away. 

The hierarchy of models is completely uninteresting to all current users of graph_walk. They're all just trying to determine the combined subscription list. Because of this, I think the simplest solution to point 1 would be to reintroduce the `next` pointer from the old approach, but use it to group all models in one extension graph together without presenting any info about the hierarchy. It could be used to create a circular linked list of all models in the graph, for instance. This'll require some linked list merging logic in the extend function, but walking would be O(n), and extending would be O(1).

Additionally, I reckon all models would know which models they extend at runtime - perhaps the model relationships could be moved from RAM to ROM by adding a model callback instead? e.g.

```c
int (*const extended_models)(struct bt_mesh_model *mod, struct bt_mesh_model **children);
```

where `children` would be an array of configurable length. The callback would need to return which models it extends in that array, and return the count. This would remove the need for the model extension graph altogether, and we can reuse information that's likely already present in the model implementation without any additional RAM. This mechanism wouldn't actually be needed at the moment though, so I think we could postpone adding this without being forced to make changes to existing applications in the future.

WilliamGFish commented:
> You were right about that graph_walk issue, that's a pretty heavy process.
> 
I have been looking at this and trying to wrap my head around the logic and if a different approach could exist.
The documentation is just not clear and find the Mesh Model Spec 1.0.1 highlights multiple models extended within section 3.3.7.1

> The Generic Power Level Setup Server model extends the Generic Power Level Server model (see Section 3.3.6) and the Generic Power OnOff Setup model (see Section 3.3.5).

The diagram on pg 104 is another example of how is should be 'defined' (Figure 3.5: Relationships between generic models)

I will have to sleep on it...

WilliamGFish commented:
Ok, I've been reading the specs and have confused myself several times interpreting the language. What I have concluded is the original tree method best reflects the intention in the documentation with one significant difference.

The LCSR Tree is upside-down:
```
	/* Form a cyclical LCRS tree:
	 * The extends-pointer points to the first child, and the next-pointer
	 * points to the next sibling. The last sibling is marked by the
	 * BT_MESH_MOD_NEXT_IS_PARENT flag, and its next-pointer points back to
	 * the parent. This way, the whole tree is accessible from any node.
	 *
	 * We add children (extend them) by inserting them as the first child.
	 */
```

If this function added the extending model as the new **_root_**  rather than a child/sibling, it would provide a better solution, IMHO.  Not sure how it would be coded as my brain is a worn out now.

Thoughts..? 

Billy...

michal-narajowski commented:
@trond-snekvik I implemented a version with an extension list as you described. It's not merge-ready but I wanted to get some comments first before I polish it.

michal-narajowski commented:
@WilliamGFish I'm not sure how that would work but it seems to me that it would also obscure relationships between individual models, so I think ditching them altogether and using a list as @trond-snekvik mentioned might be easier to grasp.

WilliamGFish commented:
> @WilliamGFish I'm not sure how that would work but it seems to me that it would also obscure relationships between individual models, so I think ditching them altogether and using a list as @trond-snekvik mentioned might be easier to grasp.

I know my suggestion seems very odd but it makes more sense if you draw the relationships.

Additionally I was having a look at how mbed implemented this but it appears they just have a list of accepted models and they are removed by commenting them out. 

Billy..

trond-snekvik reviewed (commented):
@WilliamGFish the extension graph needs to support models that extend and are extended by multiple models at the same time. For instance, the Light Lightness Server model is extended both by a Light Lightness Setup Server and by a Light CTL Server, while also extending a Power OnOff Server and a Generic Level Server. In other words, the tree is really a DAG, where each node may have multiple outgoing and incoming edges, and we need to be able to access all nodes in the graph from any other node. I believe this can't be achieved by the LCRS tree - regardless of direction, and as far as I can tell, we'd need to allocate arrays for incoming and outgoing extensions, as in the original version of this PR. If there's a variant of the tree structure that can support a full DAG without dynamic allocation, that'll be a very good candidate for a solution though.

Since these relationships aren't actually used at the moment, I think the circular linked list approach is a good solution, as it covers all current usage scenarios and configurations with optimal processing complexity for the most frequently encountered scenario, the subscription list check. It also doesn't put any memory calculation requirements on the application developer.

Commit:
Bluetooth: Mesh: Refactor Mesh Model Extensions

The existing extension tree does not support all the features that are
defined by the specification (e.g. multiple parents).

This patch approaches this problem by defining a circular single-linked
list of extension models. So for a given model, all models that are on
the same list as that model are in some extension relationship with that
model. All models on a list represent a single connected component of an
extension graph but without defining specific relationships between each
pair of models. This list is used to manage a shared subscription list
as per the Mesh Profile Specification:

```4.2.4 Subscription List

Within an element, each model has a separate instance of a Subscription
List, unless the model extends another model on that element. Instances
of models that extend other models (i.e., all models within an extension
relation tree) shall share a single instance of a Subscription List per
element.
```

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

trond-snekvik commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -812,80 +810,71 @@ const struct bt_mesh_comp *bt_mesh_comp_get(void)
-	if (base_mod->next) {
-		return -EALREADY;
+	struct bt_mesh_model *a = extending_mod, *b = base_mod, *a_next = a->next,
+			     *b_next = b->next, *it;
```
Nitpicking: From the [Linux Kernel codingg style page, part 8](https://www.kernel.org/doc/html/v4.10/process/coding-style.html#commenting):

> use just one data declaration per line (no commas for multiple data declarations). This leaves you room for a small comment on each item, explaining its use.

trond-snekvik commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -812,80 +810,71 @@ const struct bt_mesh_comp *bt_mesh_comp_get(void)
+	}
+
+	/* Check if a's list contains b */
+	for (it = a; it && it->next != a; it = it->next) {
```
Also needs a check for `a->next == NULL` or `it != NULL`
